### PR TITLE
Allow buffered copy

### PR DESCRIPTION
**Pull request by guenhter**
Add an option to define a buffer size when it comes to copying files.

### CODE REVIEW

File: copy.go [hunk 1]
```diff
@@ -87,7 +89,13 @@ func fcopy(src, dest string, info os.FileInfo, opt Options) (err error) {
 	}
 	defer fclose(s, &err)
 
-	if _, err = io.Copy(f, s); err != nil {
+	if opt.CopyBufferSize <= 0 {
+		_, err = io.Copy(f, s)
+	} else {
+		buf := make([]byte, opt.CopyBufferSize)
+		_, err = io.CopyBuffer(onlyWriter{f}, s, buf)
```
Comment (guenhter) [inline copy.go:96]: Wrapping `onlyWriter{f}` is necessary, because else the specified buffer is never used internally, because the file writer complies to the `io.ReaderFrom` interface which has the effect, that the `CopyBuffer` function not used the buffer but instead forwards the request to the `io.ReaderFrom`. This `ReaderFrom` is sometimes a good choice but in case it can't handle the request (e.g. too old kernel) it falls back to the previous copy function but without any buffer (except the default 32K): See


Excerpt of `file.go`
```
// ReadFrom implements io.ReaderFrom.
func (f *File) ReadFrom(r io.Reader) (n int64, err error) {
	if err := f.checkValid("write"); err != nil {
		return 0, err
	}
	n, handled, e := f.readFrom(r)
	if !handled {
		return genericReadFrom(f, r) // without wrapping
	}
	return n, f.wrapErr("write", e)
}
```

**Comment by guenhter** [comment]
@otiai10 fixed problems of the CI. Mainly compatibility issues of the test with older go versions.

File: all_test.go [hunk 2]
```diff
@@ -38,8 +37,11 @@ func TestCopy(t *testing.T) {
 	})
 
 	When(t, "specified src is just a file", func(t *testing.T) {
-		err := Copy("test/data/case01/README.md", "test/data.copy/case01/README.md")
+		err := Copy("test/data/case00/README.md", "test/data.copy/case00/README.md")
```
Comment (otiai10) [inline all_test.go:40]: Why did you change case01 to case00 specifically here?

File: all_test.go [hunk 3]
```diff
@@ -242,6 +244,19 @@ func TestOptions_PreserveTimes(t *testing.T) {
 	}
 }
 
+func TestOptions_CopyBufferSize(t *testing.T) {
+	opt := Options{
+		CopyBufferSize: 512,
+	}
+
+	err := Copy("test/data/case00", "test/data.copy/case00", opt)
```
Comment (otiai10) [inline all_test.go:252]: Just for naming: Why don't you append a test case, rather than using `00`?

File: copy.go [hunk 4]
```diff
@@ -43,22 +47,20 @@ func switchboard(src, dest string, info os.FileInfo, opt Options) (err error) {
 	default:
 		err = fcopy(src, dest, info, opt)
 	}
-
 	return err
 }
 
-// copy decide if this src should be copied or not.
+// copySkipping decide if this src should be copied or not.
```
Comment (otiai10) [inline copy.go:53]: Do we really need to change the name of this internal method? and the name `copySkipping` doesn't make sense to me. I prefer to keep just `copy` here. Let me know your thought.

File: copy.go [hunk 5]
```diff
@@ -87,7 +89,13 @@ func fcopy(src, dest string, info os.FileInfo, opt Options) (err error) {
 	}
 	defer fclose(s, &err)
 
-	if _, err = io.Copy(f, s); err != nil {
+	if opt.CopyBufferSize <= 0 {
+		_, err = io.Copy(f, s)
+	} else {
+		buf := make([]byte, opt.CopyBufferSize)
```
Comment (otiai10) [inline copy.go:95]: Since `io.Copy` uses `copyBuffer` internally, I don't think using `if opt.CopyBufferSize <= 0` is a simple implementation.

Just for an idea: How about using default option with size `32 * 1024` used by [`io.Copy`](https://golang.org/src/io/io.go?s=14517:14534#L412) so we don't need to use `if` here.

**Review by otiai10 — COMMENTED** [review]
Thank you for your proposal.
I added some comments. I'd love to know your opinion on them.

File: all_test.go (hunk 2 above)
Comment (guenhter) [reply all_test.go:40]: I added some content to the case00/README.md and check then the content of the file as well. I can revert this also if you want.

File: all_test.go (hunk 3 above)
Comment (guenhter) [reply all_test.go:252]: Oh, I didn't see that the "caseXX" files match the actual test steps. I'll change that. Thx for letting me know.

File: copy.go (hunk 4 above)
Comment (guenhter) [reply copy.go:53]: I only changed it, because `copy` itself is a go-keyword and if possible should not be used to prevent any confusion. Would you have a better name for that function, then I'd change it accordingly.

File: copy.go (hunk 5 above)
Comment (guenhter) [reply copy.go:95]: Perfectly fine. I'll change this.

Comment (guenhter) [reply copy.go:95]: I changed this a little bit, but I need at least one `if` to check if the bufferSize was set appropriately

File: copy.go (hunk 4 above)
Comment (guenhter) [reply copy.go:53]: I reverted this to `copy` but you really should think about a different name to avoid confusions.

File: all_test.go (hunk 3 above)
Comment (otiai10) [reply all_test.go]: When we add some features, it should be another test case to be appended.

File: all_test.go (hunk 2 above)
Comment (otiai10) [reply all_test.go]: This is a test case for "the src is just a file", and I use `case01` for this case. So if you think this case should contain some contents, you can add some text into `case01`.

Comment (guenhter) [reply all_test.go]: Oh I see the point of the `casexx` files now. I reverted this.

File: all_test.go (hunk 3 above)
Comment (guenhter) [reply all_test.go]: Thx. Done.

File: copy.go (hunk 4 above)
Comment (otiai10) [reply copy.go]: Makes sense. Let me think.

Comment (otiai10) [reply copy.go]: I'm gonna change the name after I merge your PR. Please just leave my comment here ;)

File: copy.go (hunk 5 above)
Comment (otiai10) [reply copy.go]: I mean using `https://github.com/otiai10/copy/blob/main/options.go#L63` and `https://github.com/otiai10/copy/blob/main/copy.go#L205` to setup default buffer size, then we don't need to judge if it's set.
If the user intentionally set 0 or negative, let it error.
(or we can just prevent negative to use `uint` here)

File: copy.go [hunk 6]
```diff
@@ -21,6 +23,10 @@ type timespec struct {
 	Ctime time.Time
 }
 
+type onlyWriter struct {
```
Comment (otiai10) [inline copy.go:26]: Sorry I couldn't give feedbacks in one time but, why do we need this type?
Can we just give `*os.File` to `CopyBuffer`?

Comment (guenhter) [reply copy.go:26]: Unfortunately, there is a "feature" in the copyBuffer which prevents that when we add a `os.File` the buffer is not used but it is bypassed to some other "optimization". That optimization is often not working (because of e.g. too old linux kernels) and then falls back to the plain old copy function but with the default buffer size.
So it is needed to wrap the file to prevent any detours and go streight to the copy command with the buffer.

File: copy.go (hunk 5 above)
Comment (guenhter) [reply copy.go]: Wouldn't that mean I break everyones code who creates the `Options{` themself (without the `getDefaultOptions` function. When they upgrade to this version, their code breaks because this option would be 0 by default...

Comment (guenhter) [reply copy.go]: Oh I see. Yes, I love that.

Comment (guenhter) [reply copy.go]: Pushed that change. I didn't had a deep enough look into your lib to find out about this approach. Very cool. Thx

File: copy.go (hunk 6 above)
Comment (guenhter) [reply copy.go:26]: If you have a look at this line https://github.com/golang/go/blob/ba0f8ce50fd982fe6e51f77c515660fffdebe40c/src/io/io.go#L409

This is exactly the bypass I'm talking about. Sometimes this is a good thing, but sometimes it isn't (like if your kernel is bellow 5.3).

Comment (otiai10) [reply copy.go:26]: OK, I've got your point. Thank you for clarification.

IMO, it's better to have another option _to use / not to use_ that optimization which depends on the env of the user.

Comment (guenhter) [reply copy.go:26]: Do you think we should add a control flag to tell if we should wrap the `os.File` or do you think we should use the `bufferSize` and use the plain `Copy` in case of bufferSize is 0 and use the `CopyBuffer` with the wrapping in case the buffersize is valid.

Comment (guenhter) [reply copy.go:26]: Because for calling the `io.CopyBuffer` function with a `os.File` as `Writer` will never ever consider the given buffer. So providing a buffer with `os.File` as reader makes actually no sense at all ... never. 
So letting the buffersize be specified AND setting a flag that we don't want to wrap the writer would be quite useless, because then the buffersize has no effect.

Comment (guenhter) [reply copy.go:26]: This would somehow bring back the old way:

```
if copyBuffer > 0 {
    ... io.CopyBuffer(...)
} else {
   .... io.Copy(...)
}
```

Comment (guenhter) [reply copy.go:26]: @otiai10 What do you think?

Comment (otiai10) [reply copy.go:26]: > So letting the buffersize be specified AND setting a flag that we don't want to wrap the writer would be quite useless, because then the buffersize has no effect.

I understand, finally. Thanks. Then here are my points.

1. Users (≠ we, developers) should be able to explicitly set bufsize, and it must be considered if set.
2. Users should be able to explicitly know the optimization by kernel will be disabled when bufsize is set.

Thus, you are right, we should implement something like you call "the old way" 👉 

I'll take over your branch and tweak a bit, if you don't mind.

Comment (guenhter) [reply copy.go:26]: Be my guest. Thank you very much. Appreciate that.

**Comment by otiai10** [comment]
@guenhter Thank you for your patience. Take a look at https://github.com/otiai10/copy/pull/61 and leave some reaction. Thank you ;)
